src/nvim/event/loop.c

```diff
@@ -138,6 +138,40 @@ static void loop_walk_cb(uv_handle_t *handle, void *arg)
 }
 #endif
 
+static bool loop_uv_close(uv_loop_t *loop, bool wait)
+{
+  // Run the loop to tickle close-callbacks (which should then free memory).
+  // Use UV_RUN_NOWAIT to avoid a hang. #11820
+  uv_run(loop, UV_RUN_NOWAIT);
+
+  if (!uv_loop_close(loop) || !wait) {
```

Comment:
why remove the `UV_EBUSY` check? that seems like a logic change, not just a refactor. should be separate.

```suggestion
  if ((uv_loop_close(loop) != UV_EBUSY) || !wait) {
```

Reply:
This function currently only returns 0 of UV_EBUSY so it is functionally equivalent. If a new error return value is added in the future then our current code will just break and return prematurely. The goal of this code is to keep running uv_close_loop until it is successful.

Reply:
> The goal of this code is to keep running uv_loop_close until it is successful.

Are you sure? This part is intended to return quickly, that's why it checks `!wait`. If some other state besides "busy" is introduced, it could make sense to return early then also.

Reply:
This diff is outdated so probably not much benefit in continuing this. I'm mostly concerned about the wait scenario anyway. The new diff returns early just on `!wait`.

Reply:
but the new diff does not check the result of `uv_loop_close`, in the case of `wait`. so that is a logic change.

Reply:
I guess so, but I'm not sure if it's significant.

Now we have:
```c
 // Run the event loop until all handles are successfully closed
  while (uv_loop_close(loop)) {
    uv_run(loop, UV_RUN_DEFAULT);
    uint64_t elapsed_s = (os_hrtime() - start) / 1000000000;  // seconds
    if (elapsed_s >= 2) {
      // Some libuv resource was not correctly deref'd. Log and bail.
      ELOG("uv_loop_close() hang?");
      log_uv_handles(loop);
      return false;
    }
  }
  return true;
```

Which is pretty similar to what luv does, and is a lot easier to follow.

I guess elapsed is checked slightly different. I can move that code above `uv_run` which makes it closer to was there before.

Reply:
I think the only difference now is that in the case of `wait`, we no longer run `uv_run(loop, UV_RUN_NOWAIT)` at the beginning.

Reply:
> I guess so, but I'm not sure if it's significant.

logic changes should be a separate PR, helps with bisecting etc.

> Which is pretty similar to what luv does,

why does that matter? Nvim has different use-cases, especially when exiting.

Reply:
> logic changes should be a separate PR, helps with bisecting etc.

If the logic ends up being different I'll change the commit message from refactor.

> why does that matter? Nvim has different use-cases, especially when exiting.

It doesn't strictly matter. I'm just not too comfortable with the logic here. More than anything this PR is an exercise for me to better understand how the loop works and attempt to make sure we are handling the loop closing as best we can.